### DISCUSSION TITLE
Help needed for xml value deserialisation

### DISCUSSION THREAD

**Discussion by frichez** [pdvrieze/xmlutil#141] [Q&A]
I want to deserialise this xml:

```
 <Tag ID="6F" name="FCI Template">
     <Tag ID="84" name="DF Name">A0 00 00 00 03 10 10</Tag> 
     <Tag ID="A5" name="FCI Proprietary Template">
         <Tag ID="50" name="Application Label">VISA CREDIT</Tag>
     </Tag>
</Tag>
```

Here is the code:

```
@Serializable
public data class Tag(
    @SerialName("ID") val id: String,
    val name: String,
    //@XmlValue(true) val content: String?,
    val tags: List<Tag>
)
```

I successfully get the Tag structure except xml values ("A0 00 00 00 03 10 10" and "VISA CREDIT")
I tried to use the @XmlValue(true) annotation to get the value but I get the exception "Types with an @XmlValue member may not contain other child elements".

Any hit how can I get those values ?

Thks.

**Comment by chumpa** [comment]
Just a remark, it's Mixed Content type of xml. 
https://github.com/pdvrieze/xmlutil declares `@XmlMixed` but I cannot find it.

**Comment by pdvrieze** [comment]
This is supported, but you'll have to look at the tests for an example: https://github.com/pdvrieze/xmlutil/blob/master/serialization/src/commonTest/kotlin/nl/adaptivity/xml/serialization/MixedValueContainerTest.kt and https://github.com/pdvrieze/xmlutil/blob/master/serialization/src/commonTest/kotlin/nl/adaptivity/xml/serialization/TestMixedTypesafe.kt
In one of the tests there is some version with xsi etc. you can ignore that, it is not needed. The examples show two versions of how this can be done.

**Reply by chumpa** [reply]
where is the annotation @XmlMixed declared?

**Comment by frichez** [comment]
Thks for the hint, here is my code, comments for improvements welcome ;-)

1/ In the policy there is a "pedantic" value, what does it means ?
2/ When I write @XmlValue(true) val data... instead of @XmlValue val data... I get an exception during compilation : "_Backend Internal error: Exception during IR lowering_", same for other xml annotations I had to specify all parameters, this is not the case in xmlutil unit test cases (I'm using 0.85.0 version).

Thks. 

```
class TagParserTest {
    @Test
    fun parse() {
        val xml = XML(Tag.module) {
            defaultPolicy {
                //pedantic = false ???
                ignoreUnknownChildren()
                autoPolymorphic = true
            }
        }
        val data = """
            <Tag ID="6F" name="FCI Template">
                <Tag ID="84" name="DF Name">A0 00 00 00 03 10 10</Tag>
                <Tag ID="A5" name="FCI Proprietary Template">
                    <Tag ID="50" name="Application Label" format="ISO-8859-1">VISA CREDIT</Tag>
                </Tag>
            </Tag>"""
        val tag = xml.decodeFromString<Tag>(data)

        assertEquals("6F", tag.id)
        assertEquals("FCI Template", tag.name)
        with(tag.children[0]) {
            assertEquals("84", id)
            assertEquals("DF Name", name)
            assertEquals("A0 00 00 00 03 10 10", value)
        }
        with(tag.children[1]) {
            assertEquals("A5", id)
            assertEquals("FCI Proprietary Template", name)
            with(children[0]) {
                assertEquals("50", id)
                assertEquals("Application Label", name)
                assertEquals("VISA CREDIT", value)
            }
        }
    }
}

@Serializable
@SerialName("Tag")
internal data class Tag(
    @SerialName("ID") val id: String,
    @SerialName("name") val name: String,
    @XmlValue(true) val data: List<MixedContent<Tag>>,
) {
    @Transient
    val children get() =  data.filterIsInstance<MixedContent.Object<Tag>>().map { it.data }
    @Transient
    val value get() = data.filterIsInstance<MixedContent.Text>().joinToString(separator = "") { it.data }.trim()

    companion object {
        val module = SerializersModule {
            polymorphic(Any::class) {
                subclass(String::class)
                subclass(Tag::class)
            }
        }
    }
}
```

**Reply by chumpa** [reply]
I am also periodically got "Backend Internal error: Exception during IR lowering". Sometimes it helps me to change kotlinx.serialization + kotlin itself version.

**Reply by pdvrieze** [reply]
Unfortunately this is a serialization plugin error that is only triggered on incremental compilation. To avoid it, don't use default parameters for the annotations (it should work, but doesn't in all cases).

**Reply by pdvrieze** [reply]
Btw. I had a quick look again. Pedantic currently is only used on initialization. It basically triggers an exception of you use invalid annotations (like trying structs to be serialized as attributes). For new code it should probably be true (as the library builds the structure for the type on construction this should be deterministic).
